Proposal.md

```diff
@@ -0,0 +1,604 @@
+//                                 point
+//
+//    ┌─────┬─────┬─────┬─────┐┌─────┐┌─────┬─────┐
+//    │  a  │  c  │  d  │  e  ││  b  ││  f  │  g  │
+//    └─────┴─────┴─────┴─────┘└─────┘└─────┴─────┘
+//                                ^
+//                        newPosition == 4
+```
+
+The new gathering and shifting methods are listed below.
+
+```swift
+extension MutableCollection {
+    /// Collects the elements at the given indices just before the specified
+    /// index.
+    ///
+    /// This example finds all the uppercase letters in the array and gathers
+    /// them between `"i"` and `"j"`.
+    ///
+    ///     var letters = Array("ABCdeFGhijkLMNOp")
+    ///     let uppercase = letters.indices(where: { $0.isUppercase })
+    ///     let rangeOfUppercase = letters.gather(uppercase, justBefore: 10)
+    ///     // String(letters) == "dehiABCFGLMNOjkp"
+    ///     // rangeOfUppercase == 4..<13
+    ///
+    /// - Parameters:
+    ///   - indices: The indices of the elements to move.
+    ///   - insertionPoint: The index to use as the destination of the elements.
+    /// - Returns: The new bounds of the moved elements.
+    ///
+    /// - Complexity: O(*n* log *n*) where *n* is the length of the collection.
+    @discardableResult
+    public mutating func gather(
+        _ indices: RangeSet<Index>, justBefore insertionPoint: Index
+    ) -> Range<Index>
+
+    /// Collects the elements that satisfy the given predicate just before the
+    /// specified index.
+    ///
+    /// This example gathers all the uppercase letters in the array between
+    /// `"i"` and `"j"`.
+    ///
+    ///     var letters = Array("ABCdeFGhijkLMNOp")
+    ///     let rangeOfUppercase = letters.gather(justBefore: 10) { $0.isUppercase }
+    ///     // String(letters) == "dehiABCFGLMNOjkp"
+    ///     // rangeOfUppercase == 4..<13
+    ///
+    /// - Parameters:
+    ///   - predicate: A closure that returns `true` for elements that should
+    ///     move to `destination`.
+    ///   - insertionPoint: The index to use as the destination of the elements.
+    /// - Returns: The new bounds of the moved elements.
+    ///
+    /// - Complexity: O(*n* log *n*) where *n* is the length of the collection.
+    @discardableResult
+    public mutating func gather(
+        justBefore: Index, where predicate: (Element) -> Bool
+    ) -> Range<Index>
+    
+    /// Shifts the elements in the given range to the specified insertion point.
+    ///
+    /// - Parameters:
+    ///   - range: The range of the elements to move.
+    ///   - insertionPoint: The index to use as the insertion point for the 
+    ///     elements. `insertionPoint` must be a valid index of the collection.
+    /// - Returns: The new bounds of the moved elements.
+    ///
+    /// - Returns: The new bounds of the moved elements.
+    /// - Complexity: O(*n*) where *n* is the length of the collection.
+    @discardableResult
+    public mutating func shift(
```

Comment:
I know we talked about this, but I don't understand why this would be called `shift` instead of `gather`.

Reply:
While `shift` is a specialization of `gather`, they're not at all the same from an end-user or call-site perspective. It doesn't make sense to "gather" a single index, and a range is already "gathered".

Reply:
I can say, "this group should gather over there," and it makes sense even if they're already gathered over here.  And I'm not convinced that the differing language is worth breaking up the continuity of naming across semantically identical things.  When you start with a contiguous range and need to switch to a discontiguous range set, should you have to change the algorithm name?  Is there a place for generic algorithms that operate on either a range or a range set, with both conforming to the same protocol?

I'm not convinced I'm definitely right, mind you, but I don't think it's open-and-shut that I'm wrong either.